added test goodfeaturetotrack.cpp

AlexanderUsentsov opened the pull request:
Added missed accuracy test for goodFeaturesToTrack functions from opencv_imgproc module

vpisarev commented:
@AlexanderUsentsov, thank you for the PR. There are still failing tests: http://pullrequest.opencv.org

AlexanderUsentsov commented:
I am working on these issues.

Thank you.
Alexander

From: Vadim Pisarevsky [mailto:notifications@github.com]
Sent: Tuesday, January 20, 2015 4:04 PM
To: Itseez/opencv
Cc: Usentsov, Alexander
Subject: Re: [opencv] added test goodfeaturetotrack.cpp (#3540)


@AlexanderUsentsov<https://github.com/AlexanderUsentsov>, thank you for the PR. There are still failing tests: http://pullrequest.opencv.org

—
Reply to this email directly or view it on GitHub<https://github.com/Itseez/opencv/pull/3540#issuecomment-70649859>.

--------------------------------------------------------------------
Closed Joint Stock Company Intel A/O
Registered legal address: Krylatsky Hills Business Park, 
17 Krylatskaya Str., Bldg 4, Moscow 121614, 
Russian Federation

This e-mail and any attachments may contain confidential material for
the sole use of the intended recipient(s). Any review or distribution
by others is strictly prohibited. If you are not the intended
recipient, please contact the sender and delete all copies.

ilya-lavrenov commented:
@AlexanderUsentsov, any progress?

AlexanderUsentsov commented:
Hello, Ilya!

I have fixed all bugs in accuracy test from opencv_test_imgproc but can’t  fix issues in CV_FeatureDetectorTest::regressionTest() from opencv_test_features2d .
I see that result is compared with some test data from ..\opencv_extra-master\testdata/cv/features2d/feature_detectors/detector-harris.xml.gz files.  As I changed code of this function  I receive different results to compare with old code. To fix this issue I need to change detector-harris.xml.gz file.
Could you help me to generate new test data file?

Thank you,
Alexander

From: Ilya Lavrenov [mailto:notifications@github.com]
Sent: Saturday, February 21, 2015 12:47 PM
To: Itseez/opencv
Cc: Usentsov, Alexander
Subject: Re: [opencv] added test goodfeaturetotrack.cpp (#3540)


@AlexanderUsentsov<https://github.com/AlexanderUsentsov>, any progress?

—
Reply to this email directly or view it on GitHub<https://github.com/Itseez/opencv/pull/3540#issuecomment-75364330>.

--------------------------------------------------------------------
Closed Joint Stock Company Intel A/O
Registered legal address: Krylatsky Hills Business Park, 
17 Krylatskaya Str., Bldg 4, Moscow 121614, 
Russian Federation

This e-mail and any attachments may contain confidential material for
the sole use of the intended recipient(s). Any review or distribution
by others is strictly prohibited. If you are not the intended
recipient, please contact the sender and delete all copies.

vpisarev commented:
@AlexanderUsentsov, please, use this guide to update the performance data:
http://code.opencv.org/projects/opencv/wiki/HowToUsePerfTests#How-to-update-perf-data

the updated data should be contributed to opencv_extra from the branch with the same name ("good_feature" in this case)

AlexanderUsentsov commented:
Thank you.

I updated the performance data, contributed to opencv_extra from the "good_feature" branch and opened new pull request from this branch. Unfortunately I don’t see start of building and testing . Where am I wrong?

BR,
Alexander


From: Vadim Pisarevsky [mailto:notifications@github.com]
Sent: Monday, March 2, 2015 2:58 PM
To: Itseez/opencv
Cc: Usentsov, Alexander
Subject: Re: [opencv] added test goodfeaturetotrack.cpp (#3540)


@AlexanderUsentsov<https://github.com/AlexanderUsentsov>, please, use this guide to update the performance data:
http://code.opencv.org/projects/opencv/wiki/HowToUsePerfTests#How-to-update-perf-data

the updated data should be contributed to opencv_extra from the branch with the same name ("good_feature" in this case)

—
Reply to this email directly or view it on GitHub<https://github.com/Itseez/opencv/pull/3540#issuecomment-76699991>.

--------------------------------------------------------------------
Closed Joint Stock Company Intel A/O
Registered legal address: Krylatsky Hills Business Park, 
17 Krylatskaya Str., Bldg 4, Moscow 121614, 
Russian Federation

This e-mail and any attachments may contain confidential material for
the sole use of the intended recipient(s). Any review or distribution
by others is strictly prohibited. If you are not the intended
recipient, please contact the sender and delete all copies.

mshabunin commented:
@AlexanderUsentsov, I've restarted the builds.

AlexanderUsentsov commented:
I’ve updated test data. Will you restart the build or I need to do something for that?

BR,
Alexander


From: Maksim Shabunin [mailto:notifications@github.com]
Sent: Tuesday, March 17, 2015 4:40 PM
To: Itseez/opencv
Cc: Usentsov, Alexander
Subject: Re: [opencv] added test goodfeaturetotrack.cpp (#3540)


@AlexanderUsentsov<https://github.com/AlexanderUsentsov>, I've restarted the builds.

—
Reply to this email directly or view it on GitHub<https://github.com/Itseez/opencv/pull/3540#issuecomment-82357117>.

--------------------------------------------------------------------
Closed Joint Stock Company Intel A/O
Registered legal address: Krylatsky Hills Business Park, 
17 Krylatskaya Str., Bldg 4, Moscow 121614, 
Russian Federation

This e-mail and any attachments may contain confidential material for
the sole use of the intended recipient(s). Any review or distribution
by others is strictly prohibited. If you are not the intended
recipient, please contact the sender and delete all copies.

mshabunin commented:
>opened new pull request from this branch

@AlexanderUsentsov, I don't see any PRs to _opencv_extra_ from you. Can you take a look?
Although your branch is used during the build, you still need to create PR in case to merge it.

AlexanderUsentsov commented:
Hi!
I’ve analyzed result of testing and see that  Features2d_Detector_Harris.regression test fails because ../opencv_extra-master/testdata/cv/features2d/feature_detectors/detector-harris.xml.gz files has data for previous variant of goodFeaturesToTrack. Could you help me to generate new test data file for this test?
BR,
Alexander


From: Maksim Shabunin [mailto:notifications@github.com]
Sent: Tuesday, March 24, 2015 6:07 PM
To: Itseez/opencv
Cc: Usentsov, Alexander
Subject: Re: [opencv] added test goodfeaturetotrack.cpp (#3540)


opened new pull request from this branch

@AlexanderUsentsov<https://github.com/AlexanderUsentsov>, I don't see any PRs to opencv_extra from you. Can you take a look?
Although your branch is used during the build, you still need to create PR in case to merge it.

—
Reply to this email directly or view it on GitHub<https://github.com/Itseez/opencv/pull/3540#issuecomment-85543552>.

--------------------------------------------------------------------
Closed Joint Stock Company Intel A/O
Registered legal address: Krylatsky Hills Business Park, 
17 Krylatskaya Str., Bldg 4, Moscow 121614, 
Russian Federation

This e-mail and any attachments may contain confidential material for
the sole use of the intended recipient(s). Any review or distribution
by others is strictly prohibited. If you are not the intended
recipient, please contact the sender and delete all copies.

AlexanderUsentsov commented:
Hi!

This is a friendly reminder.

BR,
Alexander

From: Usentsov, Alexander
Sent: Tuesday, April 7, 2015 7:21 PM
To: Itseez/opencv; Itseez/opencv
Subject: RE: [opencv] added test goodfeaturetotrack.cpp (#3540)

Hi!
I’ve analyzed result of testing and see that  Features2d_Detector_Harris.regression test fails because ../opencv_extra-master/testdata/cv/features2d/feature_detectors/detector-harris.xml.gz files has data for previous variant of goodFeaturesToTrack. Could you help me to generate new test data file for this test?
BR,
Alexander


From: Maksim Shabunin [mailto:notifications@github.com]
Sent: Tuesday, March 24, 2015 6:07 PM
To: Itseez/opencv
Cc: Usentsov, Alexander
Subject: Re: [opencv] added test goodfeaturetotrack.cpp (#3540)


opened new pull request from this branch

@AlexanderUsentsov<https://github.com/AlexanderUsentsov>, I don't see any PRs to opencv_extra from you. Can you take a look?
Although your branch is used during the build, you still need to create PR in case to merge it.

—
Reply to this email directly or view it on GitHub<https://github.com/Itseez/opencv/pull/3540#issuecomment-85543552>.

--------------------------------------------------------------------
Closed Joint Stock Company Intel A/O
Registered legal address: Krylatsky Hills Business Park, 
17 Krylatskaya Str., Bldg 4, Moscow 121614, 
Russian Federation

This e-mail and any attachments may contain confidential material for
the sole use of the intended recipient(s). Any review or distribution
by others is strictly prohibited. If you are not the intended
recipient, please contact the sender and delete all copies.

AlexanderUsentsov commented:
Hi!
I’ve update  opencv_extra and sent PR.  Unfortunately testing is not start.

Could you describe the procedure how to start testing after update testdata?

BR,
Alexander



From: Maksim Shabunin [mailto:notifications@github.com]
Sent: Tuesday, March 24, 2015 6:07 PM
To: Itseez/opencv
Cc: Usentsov, Alexander
Subject: Re: [opencv] added test goodfeaturetotrack.cpp (#3540)


opened new pull request from this branch

@AlexanderUsentsov<https://github.com/AlexanderUsentsov>, I don't see any PRs to opencv_extra from you. Can you take a look?
Although your branch is used during the build, you still need to create PR in case to merge it.

—
Reply to this email directly or view it on GitHub<https://github.com/Itseez/opencv/pull/3540#issuecomment-85543552>.

--------------------------------------------------------------------
Closed Joint Stock Company Intel A/O
Registered legal address: Krylatsky Hills Business Park, 
17 Krylatskaya Str., Bldg 4, Moscow 121614, 
Russian Federation

This e-mail and any attachments may contain confidential material for
the sole use of the intended recipient(s). Any review or distribution
by others is strictly prohibited. If you are not the intended
recipient, please contact the sender and delete all copies.

mshabunin commented:
@AlexanderUsentsov, I still can not find your PR to _opencv_extra_: [here](https://github.com/Itseez/opencv_extra/pulls). However, your branch is actually used for testing, please, see last build logs: [linux-build-3044](http://pullrequest.opencv.org/buildbot/builders/precommit_linux64/builds/3044) -> [merge opencv_extra test branch](http://pullrequest.opencv.org/buildbot/builders/precommit_linux64/builds/3044/steps/Merge%20opencv_extra%20with%20test%20branch/logs/stdio)

As for generating test data, as I can see in [the test](https://github.com/Itseez/opencv/blob/master/modules/features2d/test/test_detectors_regression.cpp#L211), you need to remove old validation file and run corresponding regression test - it should create the new file in this case.

AlexanderUsentsov commented:
I have two branches with name good_feature : first one for  opencv, and second  for opencv_extra. When I submit code in opencv good_feature branch testing is started. You see this result for my commit on Friday.
I set new PR when I submitted new  open_extra data today but tests didn’t start.

About generating new test data for regression test.
I deleted  detector-harris.xml.gz file and ran opencv_test_features2dd.exe --gtest_filter=*Features2d_Detector_Harris.regression*
New file was created but it has size 1 KB instead of 3KB. And test cannot load data from it.
If I run opencv_test_features2dd.exe I have the same result.

Thank you,
Alexander



From: Maksim Shabunin [mailto:notifications@github.com]
Sent: Monday, April 20, 2015 5:52 PM
To: Itseez/opencv
Cc: Usentsov, Alexander
Subject: Re: [opencv] added test goodfeaturetotrack.cpp (#3540)


@AlexanderUsentsov<https://github.com/AlexanderUsentsov>, I still can not find your PR to opencv_extra: here<https://github.com/Itseez/opencv_extra/pulls>. However, your branch is actually used for testing, please, see last build logs: linux-build-3044<http://pullrequest.opencv.org/buildbot/builders/precommit_linux64/builds/3044> -> merge opencv_extra test branch<http://pullrequest.opencv.org/buildbot/builders/precommit_linux64/builds/3044/steps/Merge%20opencv_extra%20with%20test%20branch/logs/stdio>

As for generating test data, as I can see in the test<https://github.com/Itseez/opencv/blob/master/modules/features2d/test/test_detectors_regression.cpp#L211>, you need to remove old validation file and run corresponding regression test - it should create the new file in this case.

—
Reply to this email directly or view it on GitHub<https://github.com/Itseez/opencv/pull/3540#issuecomment-94473550>.

--------------------------------------------------------------------
Closed Joint Stock Company Intel A/O
Registered legal address: Krylatsky Hills Business Park, 
17 Krylatskaya Str., Bldg 4, Moscow 121614, 
Russian Federation

This e-mail and any attachments may contain confidential material for
the sole use of the intended recipient(s). Any review or distribution
by others is strictly prohibited. If you are not the intended
recipient, please contact the sender and delete all copies.

mshabunin commented:
@AlexanderUsentsov, buildbot service does not check for changes in _opencv_extra_, thus it does not start new build when new commit arrives into this repository. You can force the build by rewriting the last commit in _opencv_ repository.
I've started new build manually.

>I have two branches with name good_feature

And you need to create two pull requests, one for _opencv_ (this one), another for _opencv_extra_.

Also, as I can see, the latest commit in your [opencv_extra/good_feature](https://github.com/AlexanderUsentsov/opencv_extra/commits/good_feature) branch was created 27 days ago. Are you sure you've pushed it today?

Can you, please, check the contents of generated file, is it xml packed in _gz_? Does it have same structure as previous?

mshabunin commented:
@AlexanderUsentsov, I was able to generate test data file without any problems. It does not have _detector_params_ section because this method is not implemented in new _GFTTDetector_ class, but keypoints information is there and the test (_Harris.regression_) passes.

AlexanderUsentsov commented:
I am able to generate test data for Harris.regression test now.

Thank you,
Alexander


From: Maksim Shabunin [mailto:notifications@github.com]
Sent: Tuesday, April 21, 2015 12:53 PM
To: Itseez/opencv
Cc: Usentsov, Alexander
Subject: Re: [opencv] added test goodfeaturetotrack.cpp (#3540)


@AlexanderUsentsov<https://github.com/AlexanderUsentsov>, I was able to generate test data file without any problems. It does not have detector_params section because this method is not implemented in new GFTTDetector class, but keypoints information is there and the test (Harris.regression) passes.

—
Reply to this email directly or view it on GitHub<https://github.com/Itseez/opencv/pull/3540#issuecomment-94724193>.

--------------------------------------------------------------------
Closed Joint Stock Company Intel A/O
Registered legal address: Krylatsky Hills Business Park, 
17 Krylatskaya Str., Bldg 4, Moscow 121614, 
Russian Federation

This e-mail and any attachments may contain confidential material for
the sole use of the intended recipient(s). Any review or distribution
by others is strictly prohibited. If you are not the intended
recipient, please contact the sender and delete all copies.

mshabunin commented:
@AlexanderUsentsov, there are some tests failures with new version:
- On MacOSX: _OCL_Imgproc/GoodFeaturesToTrack.EmptyCorners_ takes more than 180 seconds and gets killed by build service
- _OCL_GoodFeaturesToTrackFixture_GoodFeaturesToTrack.GoodFeaturesToTrack_ perf tests failed
- _OCL_PyrLKOpticalFlowFixture_PyrLKOpticalFlow.PyrLKOpticalFlow_ perf tests failed

alalek commented on modules/imgproc/src/featureselect.cpp:
```diff
@@ -305,15 +299,36 @@ void cv::goodFeaturesToTrack( InputArray _image, OutputArray _corners,
         {
             float val = eig_data[x];
-            if( val != 0 && val == tmp_data[x] && (!mask_data || mask_data[x]) )
+            if( val != 0 && (fabs(val - tmp_data[x])<CV_GFTTEPS) && (!mask_data || mask_data[x]) )
```
Please revert this change. Float equality operation is fine here.

AlexanderUsentsov replied:
Float equality operation is fine here but it is not correct in general. I propose  to use this approach.


From: Alexander Alekhin [mailto:notifications@github.com]
Sent: Monday, August 10, 2015 4:23 PM
To: Itseez/opencv
Cc: Usentsov, Alexander
Subject: Re: [opencv] added test goodfeaturetotrack.cpp (#3540)


In modules/imgproc/src/featureselect.cpp<https://github.com/Itseez/opencv/pull/3540#discussion_r36629465>:

> @@ -305,15 +299,36 @@ void cv::goodFeaturesToTrack( InputArray _image, OutputArray _corners,

>          for( int x = 1; x < imgsize.width - 1; x++ )

>          {

>              float val = eig_data[x];

> -            if( val != 0 && val == tmp_data[x] && (!mask_data || mask_data[x]) )

> +            if( val != 0 && (fabs(val - tmp_data[x])<CV_GFTTEPS) && (!mask_data || mask_data[x]) )

Please revert this change. Float equality operation is fine here.

—
Reply to this email directly or view it on GitHub<https://github.com/Itseez/opencv/pull/3540/files#r36629465>.

--------------------------------------------------------------------
Closed Joint Stock Company Intel A/O
Registered legal address: Krylatsky Hills Business Park, 
17 Krylatskaya Str., Bldg 4, Moscow 121614, 
Russian Federation

This e-mail and any attachments may contain confidential material for
the sole use of the intended recipient(s). Any review or distribution
by others is strictly prohibited. If you are not the intended
recipient, please contact the sender and delete all copies.

alalek commented on modules/imgproc/src/featureselect.cpp:
```diff
@@ -76,9 +70,9 @@ static bool ocl_goodFeaturesToTrack( InputArray _image, OutputArray _corners,
     UMat eig, maxEigenValue;
     if( useHarrisDetector )
-        cornerHarris( _image, eig, blockSize, 3, harrisK );
+        cornerHarris( _image, eig, blockSize, 3, harrisK, BORDER_REPLICATE );
```
Please revert border type to default here and below.

alalek commented on modules/imgproc/src/featureselect.cpp:
```diff
@@ -305,15 +299,36 @@ void cv::goodFeaturesToTrack( InputArray _image, OutputArray _corners,
+               pTmp = vec[i]; vec[i] = vec[j];vec[j] = pTmp;
+           }
+       }
+    }
```
`std::sort` above should work fine. Revert this change.

AlexanderUsentsov replied:
Hello!

I have finished working with code for this (#3540) pull request. All test passed.
Could you review my code?

Thanks,
Alexander


From: Alexander Alekhin [mailto:notifications@github.com]
Sent: Monday, August 10, 2015 4:26 PM
To: Itseez/opencv
Cc: Usentsov, Alexander
Subject: Re: [opencv] added test goodfeaturetotrack.cpp (#3540)


In modules/imgproc/src/featureselect.cpp<https://github.com/Itseez/opencv/pull/3540#discussion_r36629704>:

> +

> +// sort tmpCorners

> +

> +    float* pTmp;

> +    float ** vec = (float **) &tmpCorners[0];

> +

> +    for(i=0;i<total-1;i++)

> +    {

> +       for(j=i+1;j<total;j++)

> +       {

> +           if(*vec[j] > *vec[i])

> +           {

> +               pTmp = vec[i]; vec[i] = vec[j];vec[j] = pTmp;

> +           }

> +       }

> +    }

std::sort above should work fine. Revert this change.

—
Reply to this email directly or view it on GitHub<https://github.com/Itseez/opencv/pull/3540/files#r36629704>.

--------------------------------------------------------------------
Joint Stock Company Intel A/O
Registered legal address: Krylatsky Hills Business Park,
17 Krylatskaya Str., Bldg 4, Moscow 121614,
Russian Federation

This e-mail and any attachments may contain confidential material for
the sole use of the intended recipient(s). Any review or distribution
by others is strictly prohibited. If you are not the intended
recipient, please contact the sender and delete all copies.

mshabunin commented on modules/imgproc/test/test_goodfeaturetotrack.cpp:
```diff
@@ -0,0 +1,498 @@
+    const static int types[] = { CV_32FC1, CV_8UC1 };
+
+    cvtest::TS& tst = *cvtest::TS::ptr();
+    src = imread(string(tst.get_data_path()) + "stereomatching/datasets/tsukuba/im2.png", IMREAD_COLOR);
```
I suggest to use image from "shared" folder.

mshabunin commented on modules/imgproc/test/test_goodfeaturetotrack.cpp:
```diff
@@ -0,0 +1,498 @@
+        TEST_MESSAGE ("\n")
+    }
+
+    double e =norm(corners, Refcorners);
```
It is better to use `cvtest::norm` here.

mshabunin commented:
@AlexanderUsentsov , I have two more comments:

- remove or rename the `good_feature` branch from your _opencv_extra_ fork, otherwise the CI system will not be able to merge this PR
- squash all commits into one before merge

vpisarev commented:
@mshabunin, are we ready to merge it?

mshabunin commented:
@vpisarev , no, not yet.

@AlexanderUsentsov , looks like the test itself is missing now. I have a backup of the previous state if you need it: [mshabunin/opencv:good_feature](https://github.com/Itseez/opencv/compare/master...mshabunin:good_feature)

AlexanderUsentsov commented:
I am sorry.
Test was really missed in last commit.

BR,
Alexander


From: Maksim Shabunin [mailto:notifications@github.com]
Sent: Monday, December 7, 2015 1:40 PM
To: Itseez/opencv
Cc: Usentsov, Alexander
Subject: Re: [opencv] added test goodfeaturetotrack.cpp (#3540)


@vpisarev<https://github.com/vpisarev> , no, not yet.

@AlexanderUsentsov<https://github.com/AlexanderUsentsov> , looks like the test itself is missing now. I have a backup of the previous state if you need it: mshabunin/opencv:good_feature<https://github.com/Itseez/opencv/compare/master...mshabunin:good_feature>

—
Reply to this email directly or view it on GitHub<https://github.com/Itseez/opencv/pull/3540#issuecomment-162478626>.

--------------------------------------------------------------------
Joint Stock Company Intel A/O
Registered legal address: Krylatsky Hills Business Park,
17 Krylatskaya Str., Bldg 4, Moscow 121614,
Russian Federation

This e-mail and any attachments may contain confidential material for
the sole use of the intended recipient(s). Any review or distribution
by others is strictly prohibited. If you are not the intended
recipient, please contact the sender and delete all copies.

mshabunin commented:
:+1;
